Source/WebCore/css/CSSPrimitiveValueMappings.h

```diff
@@ -50,6 +50,83 @@
 
 namespace WebCore {
 
+class ImplicitValueMapper {
+public:
+    ImplicitValueMapper(const CSSValue& value)
+        : m_value(value)
+    { }
+
+    template<typename TargetType> operator TargetType() const
+    {
+        return fromCSSValueID<TargetType>(downcast<CSSPrimitiveValue>(m_value).valueID());
+    }
+
+    operator const CSSPrimitiveValue&() const
+    {
+        return downcast<CSSPrimitiveValue>(m_value);
+    }
+
+    operator unsigned short() const
+    {
+        return numericValue().value<unsigned short>();
+    }
+
+    operator int() const
+    {
+        return numericValue().value<int>();
+    }
+
+    operator unsigned() const
+    {
+        return numericValue().value<unsigned>();
+    }
+
+    operator float() const
+    {
+        return numericValue().value<float>();
+    }
+
+    operator double() const
+    {
+        return numericValue().doubleValue();
+    }
+
+private:
+    const CSSPrimitiveValue& numericValue() const
+    {
+        auto& value = downcast<CSSPrimitiveValue>(m_value);
+        ASSERT(value.primitiveType() == CSSUnitType::CSS_NUMBER || value.primitiveType() == CSSUnitType::CSS_INTEGER);
+        return value;
+    }
+
+    const CSSValue& m_value;
+};
+
+class ImplicitValueIDMapper {
+public:
+    ImplicitValueIDMapper(CSSValueID valueID)
+        : m_valueID(valueID)
+    { }
+
+    template<typename TargetType> operator TargetType() const
+    {
+        return fromCSSValueID<TargetType>(m_valueID);
+    }
+
+private:
+    CSSValueID m_valueID;
+};
+
+inline ImplicitValueMapper implicitMapper(const CSSValue& value)
+{
+    return value;
+}
+
+inline ImplicitValueIDMapper implicitMapper(CSSValueID value)
```

Comment:
Not a huge fan of the name `implicitMapper`. I think it should be named what it actually maps to, or `mapTo[...]`.

Reply:
I understand that you don’t like the name, and I am not all that fond of it myself, and open to changing it. But I can’t change it to name what it maps to, because the entire point of this is that it maps to whatever type you pass it to. That’s the feature that I am taking out of `CSSPrimitiveValue` and moving into this mapper class. So it‘s more like "map automatically" or "map based on what type you assign this to". I would absolutely be willing to rename this something else, but the suggestions you made would be for something that maps to a specific type, not automatically to whatever it’s assigned to.

Reply:
I’ll land and we can globally replace to a better name: `mapAutomatically`, `mapImplicitly`, `convertAutomatically`, `convertImplicitly`. I also think we should use this as little as possible. It’s important mainly for automatically generated code, but we could also just remove this if someone can convert all that code to not need it.

Reply:
We have `fromCSSValueID<>`, so maybe we could add `fromCSSValue<>` and this could be named `fromCSSValueDeducingType`?

Reply:
Yes, those names all seem better than implicitMapper.